File: backend/src/main/java/org/luxons/sevenwonders/game/effects/Discount.java
package org.luxons.sevenwonders.game.effects;

import java.util.ArrayList;
import java.util.List;

import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.resources.Provider;
import org.luxons.sevenwonders.game.resources.TradingRules;
import org.luxons.sevenwonders.game.resources.ResourceType;

public class Discount extends InstantOwnBoardEffect {

    private final List<ResourceType> resourceTypes = new ArrayList<>();

    private final List<Provider> providers = new ArrayList<>();

    private int discountedPrice = 1;

    public List<ResourceType> getResourceTypes() {
        return resourceTypes;
    }

    public List<Provider> getProviders() {
        return providers;
    }

    public int getDiscountedPrice() {
        return discountedPrice;
    }

    public void setDiscountedPrice(int discountedPrice) {
        this.discountedPrice = discountedPrice;
    }

    @Override
    public void apply(Board board) {
        TradingRules rules = board.getTradingRules();
        for (ResourceType type : resourceTypes) {
            for (Provider provider : providers) {
                rules.setCost(type, provider, discountedPrice);
            }
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/effects/Discount.kt
package org.luxons.sevenwonders.game.effects

import org.luxons.sevenwonders.game.boards.Board
import org.luxons.sevenwonders.game.resources.Provider
import org.luxons.sevenwonders.game.resources.ResourceType

data class Discount(
    val resourceTypes: List<ResourceType> = emptyList(),
    val providers: List<Provider> = emptyList(),
    val discountedPrice: Int = 1
) : InstantOwnBoardEffect() {

    public override fun applyTo(board: Board) {
        val rules = board.tradingRules
        for (type in resourceTypes) {
            providers.forEach { rules.setCost(type, it, discountedPrice) }
        }
    }
}
